File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/WorldStep/Data/Zone/Region/Portals/Portal/To/To.java
package ro.anud.xml_xsd.implementation.model.WorldStep.Data.Zone.Region.Portals.Portal.To;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.LocalLogger.logEnter;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturn;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturnVoid;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class To implements  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static String nodeName = "to";
    public static To fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      var instance = new To();
      if(Objects.nonNull(parent)) {
        instance.parentNode(parent);
      }
      instance.rawNode(rawNode);
      instance.deserialize(rawNode);
      return logReturn(instance);
    }
    public static To fromRawNode(RawNode rawNode) {
      logEnter();
      var instance = fromRawNode(rawNode, null);
      return logReturn(instance);
    }
    public static Optional<To> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        logEnter();
        return logReturn(rawNode.map(o -> To.fromRawNode(o, parent)));
    }
    public static List<To> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      List<To> returnList = Optional.ofNullable(rawNodeList)
          .orElse(List.of())
          .stream()
          .map(o -> To.fromRawNode(o, parent))
          .collect(Collectors.toList());
      return logReturn(returnList);
    }

    public String classTypeId() {
      return ".world_step.data.zone.region.portals.portal.to";
    }

    //Attributes

    private String zoneRef;

    private String regionRef;

    private String side;

    private Double position;
    @Builder.Default
    private Optional<String> maxWidth = Optional.empty();

    //Children elements

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private RawNode rawNode = new RawNode();

    public RawNode rawNode() {
      return rawNode;
    }
    public void rawNode(RawNode rawNode) {
      this.rawNode = rawNode;
    }

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode() {
      return parentNode;
    }

    @Builder.Default
    private List<Consumer<List<Object>>> onChangeList = new ArrayList<>();

    public String nodeName() {
      return "to";
    }

    public void notifyChange(List<Object> list) {
      var logger = logEnter();
      list.addLast(this);
      logger.log("Notify change for", this.buildPath());
      onChangeList.forEach(consumer -> consumer.accept(list));
      parentNode.ifPresent(linkedNode -> linkedNode.notifyChange(list));
    }

    public void parentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode = Optional.of(linkedNode);
      notifyChange();
    }

    public Optional<ro.anud.xml_xsd.implementation.model.WorldStep.Data.Zone.Region.Portals.Portal.Portal> parentAsPortal() {
      return parentNode.flatMap(node -> {
       if (node instanceof ro.anud.xml_xsd.implementation.model.WorldStep.Data.Zone.Region.Portals.Portal.Portal casted){
         return Optional.of(casted);
       }
       return Optional.empty();
     });
    }

    public void removeChild(Object object) {
    }

    public int buildIndexForChild(Object object) {
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(Consumer<List<Object>> onChange) {
      logEnter();
      onChangeList.add(onChange);
      return logReturn(() -> onChangeList.remove(onChange));
    }

    public void deserialize (RawNode rawNode) {
      try {
        var logger = logEnter();
        this.rawNode = rawNode;
        // Godot.GD.Print("Deserializing to");

        var innerLogger = logger.log("attributes");
        //Deserialize attributes
        innerLogger.log("zone_ref");
        this.zoneRef = rawNode.getAttributeRequired("zone_ref");
        innerLogger.log("region_ref");
        this.regionRef = rawNode.getAttributeRequired("region_ref");
        innerLogger.log("side");
        this.side = rawNode.getAttributeRequired("side");
        innerLogger.log("position");
        this.position = rawNode.getAttributeDoubleRequired("position");
        innerLogger.log("maxWidth");
        this.maxWidth = rawNode.getAttribute("maxWidth");
        innerLogger = logger.log("children");
        //Deserialize children
        logReturnVoid();
      } catch (Exception e) {
        throw new RuntimeException("Deserialization failed for: " + this.buildPath(), e);
      }
    }

    public RawNode serializeIntoRawNode()
    {
      var logger = logEnter();
      rawNode.setTag("to");
      var innerLogger = logger.log("attributes");
      //Serialize attributes
      innerLogger.log("zone_ref");
      rawNode.setAttribute("zone_ref", this.zoneRef);
      innerLogger.log("region_ref");
      rawNode.setAttribute("region_ref", this.regionRef);
      innerLogger.log("side");
      rawNode.setAttribute("side", this.side);
      innerLogger.log("position");
      rawNode.setAttribute("position", this.position);
      innerLogger.log("maxWidth");
      this.maxWidth.ifPresent(o -> rawNode.setAttribute("maxWidth", o));

      innerLogger = logger.log("children");
      //Serialize children
      return rawNode;
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing to");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }

    public String getZoneRef()
    {
      return this.zoneRef;
    }
    public To setZoneRef(String value)
    {
      this.zoneRef = value;
      notifyChange();
      return this;
    }
    public String getRegionRef()
    {
      return this.regionRef;
    }
    public To setRegionRef(String value)
    {
      this.regionRef = value;
      notifyChange();
      return this;
    }
    public String getSide()
    {
      return this.side;
    }
    public To setSide(String value)
    {
      this.side = value;
      notifyChange();
      return this;
    }
    public Double getPosition()
    {
      return this.position;
    }
    public To setPosition(Double value)
    {
      this.position = value;
      notifyChange();
      return this;
    }
    public Optional<String> getMaxWidth()
    {
      return this.maxWidth;
    }
    public To setMaxWidth(Optional<String> value)
    {
      this.maxWidth = value;
      notifyChange();
      return this;
    }

    public ro.anud.xml_xsd.implementation.util.LinkedNode deserializeAtPath(String xpath, RawNode rawNode) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }

        deserialize(rawNode);
        return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> getNodeAtPath(String xpath) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }
        return Optional.of(this);
    }
  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "value": {},
        "isSingle": true,
        "isNullable": true,
        "attributes": {
          "metaType": "object",
          "value": {
            "zone_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "region_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "side": {
              "metaType": "primitive",
              "value": "type__rectangle_side",
              "isNullable": false
            },
            "position": {
              "metaType": "primitive",
              "value": "xs:decimal",
              "isNullable": false
            },
            "maxWidth": {
              "metaType": "primitive",
              "value": "xs:positiveInteger",
              "isNullable": true
            }
          }
        }
      },
      "name": "to"
    }
  */

File: gui-client/world_step/world_step/data/nodes/node/portals/portal/to/to.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Nworld_step.Ndata.Nnodes.Nnode.Nportals.Nportal.Nto {}
namespace XSD {
}
namespace XSD.Nworld_step.Ndata.Nnodes.Nnode.Nportals.Nportal {
  public class to : IEquatable<to>, XSD.ILinkedNode  {

    public static string ClassTypeId = ".world_step.data.nodes.node.portals.portal.to";
    public static string TagName = "to";

    public string NodeName {get =>"to";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<to>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes
    private System.String _node_ref;
    public System.String node_ref { get => _node_ref; set => _node_ref = value; }
    private System.String? _side;
    public System.String? side { get => _side; set => _side = value; }
    private System.Int32 _position;
    public System.Int32 position { get => _position; set => _position = value; }
    private System.String? _maxWidth;
    public System.String? maxWidth { get => _maxWidth; set => _maxWidth = value; }

    //Children elements
    public to()
    {
    }

    public to(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public to(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
      if(name == "node_ref")
      {
        Set_node_ref(value);
      }
      if(name == "side")
      {
        Set_side(value);
      }
      if(name == "position")
      {
        Set_position(value?.ToInt() ?? 0);
      }
      if(name == "maxWidth")
      {
        Set_maxWidth(value);
      }
    }

    public void SetChild(dynamic linkedNode)
    {
    }

    public void ClearChild(dynamic linkedNode)
    {
    }

    public Action OnSelfChange(Action<to> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing to");
      //Deserialize arguments
      if(rawNode.attributes.ContainsKey("node_ref"))
      {
        var attribute_node_ref = rawNode.attributes["node_ref"];
        this.node_ref = rawNode.attributes["node_ref"];
      }
      if(rawNode.attributes.ContainsKey("side"))
      {
        var attribute_side = rawNode.attributes["side"];
        this.side = rawNode.attributes["side"];
      }
      if(rawNode.attributes.ContainsKey("position"))
      {
        var attribute_position = rawNode.attributes["position"];
        this.position = rawNode.attributes["position"];
      }
      if(rawNode.attributes.ContainsKey("maxWidth"))
      {
        var attribute_maxWidth = rawNode.attributes["maxWidth"];
        this.maxWidth = rawNode.attributes["maxWidth"];
      }

      //Deserialize children
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments
      if(this._node_ref != null)
      {
        rawNode.attributes["node_ref"] = this._node_ref.ToString();
      }
      if(this._side != null)
      {
        rawNode.attributes["side"] = this._side?.ToString();
      }
      if(this._position != null)
      {
        rawNode.attributes["position"] = this._position.ToString();
      }
      if(this._maxWidth != null)
      {
        rawNode.attributes["maxWidth"] = this._maxWidth?.ToString();
      }

      //Serialize children
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing to");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_node_ref()
    {
      return this.node_ref;
    }
    public void Set_node_ref(System.String value)
    {
      this.node_ref = value;
      this.NotifyChange();
    }
    public System.String? Get_side()
    {
      return this.side;
    }
    public void Set_side(System.String? value)
    {
      this.side = value;
      this.NotifyChange();
    }
    public System.Int32 Get_position()
    {
      return this.position;
    }
    public void Set_position(System.Int32 value)
    {
      this.position = value;
      this.NotifyChange();
    }
    public System.String? Get_maxWidth()
    {
      return this.maxWidth;
    }
    public void Set_maxWidth(System.String? value)
    {
      this.maxWidth = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      return null;
    }

    public bool IsValidChildType(ILinkedNode candidateChild) {
      return false;
    }

    public bool Equals(to? obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;

        var other = (to)obj;
        return Equals(node_ref, other.node_ref) && Equals(side, other.side) && Equals(position, other.position) && Equals(maxWidth, other.maxWidth);
    }

    public override int GetHashCode()
    {
        var acc = 0;

        acc = HashCode.Combine(acc, node_ref);
        acc = HashCode.Combine(acc, side);
        acc = HashCode.Combine(acc, position);
        acc = HashCode.Combine(acc, maxWidth);
        return acc;
    }
  }
}